File: proton-j/src/main/java/org/apache/qpid/proton/amqp/transport/Disposition.java

/*
*
* Licensed to the Apache Software Foundation (ASF) under one
* or more contributor license agreements.  See the NOTICE file
* distributed with this work for additional information
* regarding copyright ownership.  The ASF licenses this file
* to you under the Apache License, Version 2.0 (the
* "License"); you may not use this file except in compliance
* with the License.  You may obtain a copy of the License at
*
*   http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing,
* software distributed under the License is distributed on an
* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
* KIND, either express or implied.  See the License for the
* specific language governing permissions and limitations
* under the License.
*
*/


package org.apache.qpid.proton.amqp.transport;

import org.apache.qpid.proton.amqp.Binary;
import org.apache.qpid.proton.amqp.UnsignedInteger;

public final class Disposition implements FrameBody
{
    private Role _role = Role.SENDER;
    private UnsignedInteger _first;
    private UnsignedInteger _last;
    private boolean _settled;
    private DeliveryState _state;
    private boolean _batchable;

    public Role getRole()
    {
        return _role;
    }

    public void setRole(Role role)
    {
        if(role == null)
        {
            throw new NullPointerException("Role cannot be null");
        }
        _role = role;
    }

    public UnsignedInteger getFirst()
    {
        return _first;
    }

    public void setFirst(UnsignedInteger first)
    {
        if( first == null )
        {
            throw new NullPointerException("the first field is mandatory");
        }

        _first = first;
    }

    public UnsignedInteger getLast()
    {
        return _last;
    }

    public void setLast(UnsignedInteger last)
    {
        _last = last;
    }

    public boolean getSettled()
    {
        return _settled;
    }

    public void setSettled(boolean settled)
    {
        _settled = settled;
    }

    public DeliveryState getState()
    {
        return _state;
    }

    public void setState(DeliveryState state)
    {
        _state = state;
    }

    public boolean getBatchable()
    {
        return _batchable;
    }

    public void setBatchable(boolean batchable)
    {
        _batchable = batchable;
    }

    public <E> void invoke(FrameBodyHandler<E> handler, Binary payload, E context)
    {
        handler.handleDisposition(this, payload, context);
    }

    @Override
    public String toString()
    {
        return "Disposition{" +
               "role=" + _role +
               ", first=" + _first +
               ", last=" + _last +
               ", settled=" + _settled +
               ", state=" + _state +
               ", batchable=" + _batchable +
               '}';
    }
}

File: ruby/lib/core/disposition.rb
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.


module Qpid::Proton

  class Disposition
    include Util::Deprecation

    # @private
    PROTON_METHOD_PREFIX = "pn_disposition"
    # @private
    extend Util::SWIGClassHelper

    # States of a message transfer
    module State
      # Message was successfully processed by the receiver
      ACCEPTED = Cproton::PN_ACCEPTED

      # Message rejected as invalid and unprocessable by the receiver.
      REJECTED = Cproton::PN_REJECTED

      # Message was not (and will not be) processed by the receiver, but may be
      # acceptable if re-delivered to another receiver
      RELEASED = Cproton::PN_RELEASED
      # Like {RELEASED}, but there are modifications (see {Tracker#modifications})
      # that must be applied to the message by the {Sender} before re-delivering it.
      MODIFIED = Cproton::PN_MODIFIED

      # Partial message data received. Only used during link recovery.
      RECEIVED = Cproton::PN_RECEIVED

      module ClassMethods
        def name_of(state) Cproton::pn_disposition_type_name(state); end
      end
      extend ClassMethods
      def self.included(klass) klass.extend ClassMethods; end
    end
    include State

    attr_reader :impl

    # @private
    def initialize(impl, local)
      deprecated self.class, Delivery
      @impl = impl
      @local = local
      @data = nil
      @condition = nil
      @annotations = nil
    end

    # @!attribute section_number
    #
    # @return [Integer] The section number of the disposition.
    #
    proton_set_get :section_number

    # @!attribute section_offset
    #
    #  @return [Integer] The section offset of the disposition.
    #
    proton_set_get :section_offset

    # @!attribute failed?
    #
    # @return [Boolean] The failed flag.
    #
    proton_set_is :failed

    # @!attribute undeliverable?
    #
    # @return [Boolean] The undeliverable flag.
    #
    proton_set_is :undeliverable

    # Sets the data for the disposition.
    #
    # @param data [Codec::Data] The data.
    #
    # @raise [AttributeError] If the disposition is remote.
    #
    def data=(data)
      raise AttributeError.new("data attribute is read-only") unless @local
      @data = data
    end

    # Returns the data for the disposition.
    #
    # @return [Codec::Data] The data.
    #
    def data
      if @local
        @data
      else
        Codec::Data.to_object(Cproton.pn_disposition_data(@impl))
      end
    end

    # Sets the annotations for the disposition.
    #
    # @param annotations [Codec::Data] The annotations.
    #
    # @raise [AttributeError] If the disposition is remote.
    #
    def annotations=(annotations)
      raise AttributeError.new("annotations attribute is read-only") unless @local
      @annotations = annotations
    end

    # Returns the annotations for the disposition.
    #
    # @return [Codec::Data] The annotations.
    #
    def annotations
      if @local
        @annotations
      else
        Codec::Data.to_object(Cproton.pn_disposition_annotations(@impl))
      end
    end

   # Sets the condition for the disposition.
    #
    # @param condition [Codec::Data] The condition.
    #
    # @raise [AttributeError] If the disposition is remote.
    #
    def condition=(condition)
      raise AttributeError.new("condition attribute is read-only") unless @local
      @condition = condition
    end

    # Returns the condition of the disposition.
    #
    # @return [Codec::Data] The condition of the disposition.
    #
    def condition
      if @local
        @condition
      else
        Condition.convert(Cproton.pn_disposition_condition(@impl))
      end
    end

  end

end
